File: brainfuck.py
import os
import sys

def usage():
    print "%s script.b" % sys.argv[0]
    sys.exit(127)





class Machine:
    def __init__(self, filename):
        self.filename = filename
        self.code = []
        self.codePos = 0
        self.tape = []
        self.tapePos = 0

        try:
            with open(filename, 'r') as fp:
                self.code = [ ord(ch) for ch in fp.read() ]
        except Exception as err:
            print err
            sys.exit(127)

    def run(self, skip = False):
        while self.tapePos >= 0 and self.codePos < len(self.code):
            if self.tapePos >= len(self.tape):
                self.tape.append(0)

            if self.code[self.codePos] == ord('['):
                self.codePos += 1
                oldPos = self.codePos
                while self.run(self.tape[self.tapePos] == 0):
                    self.codePos = oldPos
            elif self.code[self.codePos] == ord(']'):
                return self.tape[self.tapePos] != 0
            elif not skip:
                ch = self.code[self.codePos]
                if ch == ord('+'):
                    if self.tape[self.tapePos] < 255:
                        self.tape[self.tapePos] += 1
                elif ch == ord('-'):
                    if self.tape[self.tapePos] < 255:
                        self.tape[self.tapePos] -= 1
                elif ch == ord('>'): self.tapePos += 1
                elif ch == ord('<'): self.tapePos -= 1
                elif ch == ord('.'): os.write(2, chr(self.tape[self.tapePos]))
                elif ch == ord(','): self.tape[self.tapePos] = ord(os.read(0, 1))

            self.codePos += 1
        return True


if __name__ == '__main__':
    args = sys.argv[1:]
    if len(args) < 1:
        usage()

    filename = args[0]
    m = Machine(filename)
    m.run()

File: brainfuck.go
package main

import (
	"bytes"
	"flag"
	"fmt"
	"io"
	"os"
	"unicode"
)

func init() {
	flag.Parse()
}

type Code struct {
	data []byte
	pos  int
}

func (c Code) Done() bool {
	return c.pos >= len(c.data)
}
func (c Code) Curr() byte {
	return c.data[c.pos]
}

var code Code

type Tape struct {
	data []byte
	pos  int
}

func (t Tape) Curr() byte {
	return t.data[t.pos]
}

func (t Tape) Push(b byte) {
	t.data = append(t.data, b)
}

func (t *Tape) Incr() {
	t.data[t.pos]++
}
func (t *Tape) Descr() {
	t.data[t.pos]--
}

var tape Tape

func usage() {
	fmt.Printf("usage: %s brainfuck.b\n", os.Args[0])
	os.Exit(127)
}

func main() {
	flag.Parse()
	if flag.NArg() < 1 {
		usage()
	}

	args := flag.Args()

	filename := args[0]

	fd, err := os.Open(filename)
	if err != nil {
		fmt.Fprintf(os.Stderr, "brainfuck: %s\n", err)
		os.Exit(127)
	}
	defer fd.Close()

	var buf bytes.Buffer
	io.Copy(&buf, fd)

	b := buf.Bytes()
	code = Code{data: b, pos: 0}
	tape = Tape{data: make([]byte, len(b)), pos: 0}

	run(false)
}

func run(skip bool) bool {
	for tape.pos >= 0 && !code.Done() {
		if tape.pos >= len(tape.data) {
			tape.Push(0)
		}

		if code.Curr() == '[' {
			code.pos++
			oldPos := code.pos
			for run(tape.Curr() == 0) {
				code.pos = oldPos
			}
		} else if code.Curr() == ']' {
			return tape.Curr() != 0
		} else if !skip {
			switch code.Curr() {
			case '+':
				if tape.Curr() < 255 {
					tape.Incr()
				}
			case '-':
				if tape.Curr() > 0 {
					tape.Descr()
				}
			case '>':
				tape.pos++
			case '<':
				tape.pos--
			case '.':
				fmt.Printf("%c", tape.Curr())
			case ',':
				b := make([]byte, 1)
				os.Stdin.Read(b)
				if unicode.IsPrint(rune(b[0])) {
					tape.data[tape.pos] = b[0]
				} else {
					fmt.Println()
				}
			}
		}

		code.pos++
	}

	return true
}
